### DISCUSSION TITLE
How to properly type columns when using createColumnHelper?

### DISCUSSION THREAD

**Discussion by ronaldruzicka** [TanStack/table#5218] [Q&A]
Hi, does anyone know how to properly type `columns` when table is extracted into it's own component? The type should be `ColumnDef` which is a union type where is `AccessorColumnDef` which is a union type of `AccessorKeyColumnDef |
AccessorFnColumnDef` ..but in the end TS is showing an error, that `accessorKeyFn` is not used, because I'm using `accessorKey`.

The `Props` type for my `Table`
```ts
type Props = {
  columns: ColumnDef<Person>[];
  data: Person[];
}

function Table({
  columns,
  data,
}: Props) {...}
```

The error:

![Screenshot 2023-12-19 at 09 43 26](https://github.com/TanStack/table/assets/20684387/90e05fdd-2235-401e-a760-396a5348267b)

Here is [CodeSandbox](https://codesandbox.io/p/devbox/accessorcolumndef-type-error-mc4hlq?file=%2Fsrc%2Fmain.tsx%3A89%2C5&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clqc1u27h00083b6i3v1lf5jg%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clqc1u27h00033b6ir31by290%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clqc1u27h00053b6i6naubabq%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clqc1u27h00073b6i2cbm8g5q%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B72.98321188453642%252C27.016788115463584%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clqc1u27h00033b6ir31by290%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqc1u27h00023b6ip1c2wyt5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A89%252C%2522startColumn%2522%253A5%252C%2522endLineNumber%2522%253A89%252C%2522endColumn%2522%253A5%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fmain.tsx%2522%257D%255D%252C%2522id%2522%253A%2522clqc1u27h00033b6ir31by290%2522%252C%2522activeTabId%2522%253A%2522clqc1u27h00023b6ip1c2wyt5%2522%257D%252C%2522clqc1u27h00073b6i2cbm8g5q%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqc1u27h00063b6iiuaipikq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clqc1u27h00073b6i2cbm8g5q%2522%252C%2522activeTabId%2522%253A%2522clqc1u27h00063b6iiuaipikq%2522%257D%252C%2522clqc1u27h00053b6i6naubabq%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqc1u27h00043b6iu1p74uic%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522id%2522%253A%2522clqc1u27h00053b6i6naubabq%2522%252C%2522activeTabId%2522%253A%2522clqc1u27h00043b6iu1p74uic%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

I don't want to use type assertion for `columns`, but I don't see any other way.
Any help appreciated🙏, thanks!

**Comment by writes-of-spring** [comment]
If you're using the columnHelper to define your columns, in your case 
```ts 
const columnHelper = createColumnHelper<Person>()
```

you can afford to be more lax with the types in your columns

```ts
type Props = {
  columns: ColumnDef<Person, any>[];
  data: Person[];
}
```

**Reply by ronaldruzicka** [reply]
Thanks that works, but also feels a bit "fishy", `any` is also an escape patch. But probably better than assertion in this case. But still would like to know how to properly do this. Because there isn't a way to know the `TValue` in `ColumnDef<TData, TValue>`, but `unknown` causes a type error.

**Reply by LinnJS** [reply]
Plus 1 on this @ronaldruzicka 

I was trying to do something like `columns: ColumnDef<TData, keyof TData>[];` It seems like we should be able to assume TValue since the TValue should always be a member of TData, right?

**Reply by connorwarme** [reply]
@ronaldruzicka and @LinnJS - did y'all make any headway on this? Or did you settle on `columns: ColumnDef<TData, any>[];`? Thanks!

**Reply by ronaldruzicka** [reply]
Hi @connorwarme, yes I went with `any`, didn't have much choice. But the custom type for the `meta` on `column` doesn't work.

**Reply by itshallrun** [reply]
any news?

**Reply by connorwarme** [reply]
@itshallrun hi Alex, the only update I have is what Ronald shared previously - we just utilized `any`.

**Comment by carlosriveroib** [comment]
I'm experiencing this issue. Is there any update on it?

**Reply by connorwarme** [reply]
Hey @carlosriveroib,

I just used what Ronald shared. Shortly afterward, my project was canceled so I didn't continue to work with tanstack table. Sorry I don't have more to offer. Best of luck!
